File: python-midonetclient/src/midonetclient/ipsec_site_connection.py
# Copyright (c) 2015 Midokura Europe SARL, All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from midonetclient import admin_state_up_mixin
from midonetclient import resource_base
from midonetclient import vendor_media_type

class IPSecSiteConnection(resource_base.ResourceBase,
                          admin_state_up_mixin.AdminStateUpMixin):

    media_type = vendor_media_type.APPLICATION_IPSEC_SITE_CONNECTION_JSON

    def __init__(self, uri, dto, auth):
        super(IPSecSiteConnection, self).__init__(uri, dto, auth)

    def get_id(self):
        return self.dto['id']

    def get_tenant_id(self):
        return self.dto['tenantId']

    def get_name(self):
        return self.dto['name']

    def get_description(self):
        return self.dto['description']

    def get_peer_address(self):
        return self.dto['peerAddress']

    def get_local_cidrs(self):
        return self.dto['localCidrs']

    def get_peer_cidrs(self):
        return self.dto['peerCidrs']

    def get_route_mode(self):
        return self.dto['routeMode']

    def get_mtu(self):
        return self.dto['mtu']

    def get_initiator(self):
        return self.dto['initiator']

    def get_auth_mode(self):
        return self.dto['authMode']

    def get_psk(self):
        return self.dto['psk']

    def get_status(self):
        return self.dto['status']

    def get_dpd_action(self):
        return self.dto['dpdAction']

    def get_dpd_interval(self):
        return self.dto['dpdInterval']

    def get_dpd_timeout(self):
        return self.dto['dpdTimeout']

    def get_ike_policy_id(self):
        return self.dto['ikePolicyId']

    def get_ipsec_policy_id(self):
        return self.dto['ipsecPolicyId']

    def name(self, name):
        self.dto['name'] = name
        return self

    def description(self, description):
        self.dto['description'] = description
        return self

    def peer_address(self, address):
        self.dto['peerAddress'] = address
        return self

    def local_cidrs(self, local_cidrs):
        self.dto['localCidrs'] = local_cidrs
        return self

    def peer_cidrs(self, peer_cirds):
        self.dto['peerCidrs'] = peer_cirds
        return self

    def route_mode(self, route_mode):
        self.dto['routeMode'] = route_mode
        return self

    def mtu(self, mtu):
        self.dto['mtu'] = mtu
        return self

    def initiator(self, initiator):
        self.dto['initiator'] = initiator
        return self

    def auth_mode(self, auth_mode):
        self.dto['authMode'] = auth_mode
        return self

    def psk(self, psk):
        self.dto['psk'] = psk
        return self

    def dpd_action(self, dpd_action):
        self.dto['dpdAction'] = dpd_action
        return self

    def dpd_interval(self, dpd_interval):
        self.dto['dpdInterval'] = dpd_interval
        return self

    def dpd_timeout(self, dpd_timeout):
        self.dto['dpdTimeout'] = dpd_timeout
        return self

    def ike_policy_id(self, ike_policy_id):
        self.dto['ikePolicyId'] = ike_policy_id
        return self

    def ipsec_policy_id(self, ipsec_policy_id):
        self.dto['ipsecPolicyId'] = ipsec_policy_id
        return self

File: midonet-cluster/src/main/java/org/midonet/cluster/rest_api/models/IPSecSiteConnection.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.rest_api.models;

import java.net.URI;
import java.util.List;
import java.util.UUID;

import javax.validation.constraints.NotNull;

import com.fasterxml.jackson.annotation.JsonIgnore;

import org.midonet.cluster.data.ZoomClass;
import org.midonet.cluster.data.ZoomEnum;
import org.midonet.cluster.data.ZoomEnumValue;
import org.midonet.cluster.data.ZoomField;
import org.midonet.cluster.models.Neutron;
import org.midonet.cluster.rest_api.ResourceUris;
import org.midonet.cluster.util.IPSubnetUtil;
import org.midonet.packets.IPSubnet;

@ZoomClass(clazz = Neutron.IPSecSiteConnection.class)
public class IPSecSiteConnection extends UriResource {

    @ZoomField(name = "id")
    public UUID id;

    @ZoomField(name = "tenant_id")
    public String tenantId;

    @ZoomField(name = "name")
    public String name;

    @ZoomField(name = "description")
    public String description;

    @ZoomField(name = "admin_state_up")
    public boolean adminStateUp = true;

    @ZoomField(name = "peer_address")
    public String peerAddress;

    @ZoomField(name = "local_cidrs", converter = IPSubnetUtil.Converter.class)
    public List<String> localCidrs;

    @ZoomField(name = "peer_cidrs", converter = IPSubnetUtil.Converter.class)
    public List<String> peerCidrs;

    @ZoomField(name = "route_mode")
    public RouteMode routeMode = RouteMode.STATIC;

    @ZoomField(name = "mtu")
    public Integer mtu = 1400;

    @ZoomField(name = "initiator")
    @NotNull
    public Initiator initiator;

    @ZoomField(name = "auth_mode")
    public AuthMode authMode = AuthMode.PSK;

    @ZoomField(name = "psk")
    @NotNull
    public String psk;

    @ZoomField(name = "status")
    public Status status;

    @ZoomField(name = "dpd_action")
    public DpdAction dpdAction;

    @ZoomField(name = "dpd_interval")
    public Integer dpdInterval;

    @ZoomField(name = "dpd_timeout")
    public Integer dpdTimeout;

    @JsonIgnore
    @ZoomField(name = "vpn_service_id")
    public UUID vpnServiceId;

    @ZoomField(name = "ike_policy_id")
    public UUID ikePolicyId;

    @ZoomField(name = "ipsec_policy_id")
    public UUID ipsecPolicyId;

    @ZoomEnum(clazz = Neutron.IPSecSiteConnection.Status.class)
    public enum Status {
        @ZoomEnumValue("ACTIVE") ACTIVE,
        @ZoomEnumValue("DOWN") DOWN,
        @ZoomEnumValue("BUILD") BUILD,
        @ZoomEnumValue("ERROR") ERROR,
        @ZoomEnumValue("PENDING_CREATE") PENDING_CREATE,
        @ZoomEnumValue("PENDING_UPDATE") PENDING_UPDATE,
        @ZoomEnumValue("PENDING_DELETE") PENDING_DELETE;
    }

    @ZoomEnum(clazz = Neutron.IPSecSiteConnection.DpdAction.class)
    public enum DpdAction {
        @ZoomEnumValue("CLEAR") CLEAR,
        @ZoomEnumValue("HOLD") HOLD,
        @ZoomEnumValue("RESTART") RESTART,
        @ZoomEnumValue("DISABLED") DISABLED,
        @ZoomEnumValue("RESTART_BY_PEER") RESTART_BY_PEER;
    }

    @ZoomEnum(clazz = Neutron.IPSecSiteConnection.AuthMode.class)
    public enum AuthMode {
        @ZoomEnumValue("PSK") PSK
    }

    @ZoomEnum(clazz = Neutron.IPSecSiteConnection.Initiator.class)
    public enum Initiator {
        @ZoomEnumValue("BI_DIRECTIONAL") BI_DIRECTIONAL,
        @ZoomEnumValue("RESPONSE_ONLY") RESPONSE_ONLY;
    }

    @ZoomEnum(clazz = Neutron.IPSecSiteConnection.RouteMode.class)
    public enum RouteMode {
        @ZoomEnumValue("STATIC") STATIC
    }

    public URI getUri() {
        return absoluteUri(ResourceUris.IPSEC_CONNECTIONS, id);
    }

    public URI getVpnService() {
        return absoluteUri(ResourceUris.VPN_SERVICES, vpnServiceId);
    }

    public URI getIkePolicy() {
        return absoluteUri(ResourceUris.IKE_POLICIES, ikePolicyId);
    }

    public URI getIpsecPolicy() {
        return absoluteUri(ResourceUris.IPSEC_POLICIES, ipsecPolicyId);
    }

    @Override
    @JsonIgnore
    public void create() {
        if (null == id) {
            id = UUID.randomUUID();
        }
    }

    @JsonIgnore
    public void create(UUID vpnServiceId) {
        create();
        this.vpnServiceId = vpnServiceId;
    }

    @JsonIgnore
    public void update(IPSecSiteConnection from) {
        id = from.id;
        vpnServiceId = from.vpnServiceId;
    }

}
